homeassistant/components/romy/config_flow.py

```diff
@@ -0,0 +1,138 @@
+"""Config flow for ROMY integration."""
+from __future__ import annotations
+
+from typing import Any
+
+import romy
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import zeroconf
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PASSWORD, CONF_UUID
+from homeassistant.data_entry_flow import FlowResult
+import homeassistant.helpers.config_validation as cv
+
+from .const import DOMAIN, LOGGER
+
+
+def _schema_with_defaults(
+    default_values: dict[str, Any] | None = None,
+    requires_password: bool = False,
+) -> vol.Schema:
+    if default_values is None:
+        default_values = {}
+    schema = {
+        vol.Required(CONF_HOST, default=default_values.get(CONF_HOST, "")): cv.string,
+    }
+
+    if requires_password:
+        schema[vol.Required(CONF_PASSWORD)] = vol.All(str, vol.Length(8))
+
+    return vol.Schema(schema)
+
+
+class RomyConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle config flow for ROMY."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self) -> None:
+        """Handle a config flow for ROMY."""
+        self.discovery_schema = None
+        self.discovery_info: dict[str, Any] = {}
+
+    async def async_step_user(
+        self, user_input: dict[str, str] | None = None
+    ) -> FlowResult:
+        """Handle the user step."""
+        errors: dict[str, str] = {}
+        data = self.discovery_schema or _schema_with_defaults()
+
+        if user_input:
+            # Save the user input and finish the setup
+            new_romy = await romy.create_romy(
+                user_input[CONF_HOST], user_input.get(CONF_PASSWORD, "")
+            )
+
+            await self.async_set_unique_id(new_romy.unique_id)
+
+            if not new_romy.is_initialized:
+                errors[CONF_HOST] = "cannot_connect"
+
+            if not new_romy.is_unlocked:
```

Comment:
It took me awhile to understand what was going on in this config flow and why the "_schema_with_defaults" flows was needed, etc, which looks a little different than other config flows I've seen.

I think this config flow should be broken into two steps, only showing the password step when needed rather than showing the same step but suddenly adding a password.  I think this is both a  simpler user experience and a simpler to follow implementation.

Similarly, why show the host again after discovery.

So, i think the steps should be:
- user enters host
- user enters password (if needed)
- create entry

Similarly on discovery, without the first step.

Reply:
Rewrote config flow with 2 new steps (unlock_http_interface, finish_config)

userconf:
![newa_userconf](https://github.com/home-assistant/core/assets/2132412/97d0f39f-9eab-41c2-8673-316578ba5219)

zeroconf:
![newa_zeroconf](https://github.com/home-assistant/core/assets/2132412/42d45398-db50-4a64-a9f6-5e9ce249f259)